### PR TITLE
Improve Docker

### PR DESCRIPTION
**Pull request by SuperSandro2000**

### CODE REVIEW

**Comment by kanishka-linux** [comment]
Docker files which are provided here are just sample files. Anyone can fork and make changes as per their preferences. If someone wants to use alpine image or ubuntu image then they can use it. I don't think PR is required for this.

About combining multiple statements into one, I think, having separate statements provides better readability. 

About `wkhtmltopdf` provided in the official repo, it was not headless when I checked last time. That's why instructions are written in the docker to directly fetch it from github release page.

**Comment by SuperSandro2000** [comment]
> Docker files which are provided here are just sample files. Anyone can fork and make changes as per their preferences.

Exactly what I did and now I want to upstream my changes. I didn't solely change preference stuff but improved the files in several ways.
Also most people probably won't change those Dockerfiles and then run into issues and get frustrated with this project.

> If someone wants to use alpine image or ubuntu image then they can use it.

Having good and sensible defaults is always a good idea. Alpine is only ~5MB and has no significant performance decrease. If I want to try your project and need to wait another 30 sec until that Ubuntu image is downloaded that is a bad experience that can easily be avoided. Also I and probably a lot more people already have the alpine variant of postgres downloaded and then we can easily save bandwidth, time and storage.

> I don't think PR is required for this.

This PR is not based around solely switching that one image from an debian base to an alpine one. If this was the only change I would agree with you but I overall improved the Docker files significantly so this change does not really matter in the grand scene of changes.

> About combining multiple statements into one, I think, having separate statements provides better readability.

Separate statements also increase the caching rate **BUT** you never ever put ``apt update`` and ``apt install`` into different layers cause otherwise you will eventually run into caching issue like installing old software or having outdated package lists.
Also it is a good idea to combine statements that thematically fit into one category or change the same file more than once into on statement. I didn't combine all ``RUN``s into but structured them in a cache efficient, logical way that is still pretty easy to understand.

> About wkhtmltopdf provided in the official repo, it was not headless when I checked last time. That's why instructions are written in the docker to directly fetch it from github release page.

Good catch. Reverting that change

**Comment by kanishka-linux** [comment]
>Having good and sensible defaults is always a good idea. Alpine is only ~5MB and has no significant performance decrease.

Yeah sure, but I develop/test my personal projects mostly on arch linux or ubuntu/debian based system. Therefore, I prefer to have docker image which resembles my development setup. Every linux distro has its own quirks and peculiarities. What works on one setup does not need to work on other setup in the same way. I always get somewhat different results trying to test project on multiple distros. Having docker environment same as that of my development environment atleast gives me assurance that things are going to work as expected without much change.

> Also I and probably a lot more people already have the alpine variant of postgres downloaded and then we can easily save bandwidth, time and storage.

That's what I was saying earlier. It is upto the user to decide what configuration to pass or not. `A lot more people` does not mean `all`, and we don't have exact statistics for that to decide. So let it leave with the end user to decide what to do. No need to send PR for this, since every person can have different preferences.

> BUT you never ever put apt update and apt install into different layers cause otherwise you will eventually run into caching issue like installing old software or having outdated package lists.

I never faced issue like this. Maybe in some rolling release distro one can face this issue, but not with LTS releases. Anyway once you do `apt update`, it updates only package list. Once you've updated package list, `apt install` will always fetch newer package as per the list, whether you put it in another layer or not. It doesn't matter. 

However, doing `apt update` without following it with `apt upgrade`, and then installing the required package, can certainly break the system.

**Comment by kanishka-linux** [comment]
If something is working really well without problems, I prefer not to change it just for the sake of change. I have not tested reminiscence inside docker for quite some time, so if you come across any bug in the docker files, you are free to point it out and submit the changes.

**Comment by SuperSandro2000** [comment]
> Yeah sure, but I develop/test my personal projects mostly on arch linux or ubuntu/debian based system. Therefore, I prefer to have docker image which resembles my development setup. Every linux distro has its own quirks and peculiarities. What works on one setup does not need to work on other setup in the same way. I always get somewhat different results trying to test project on multiple distros. Having docker environment same as that of my development environment atleast gives me assurance that things are going to work as expected without much change.

Only the postgres is on alpine and all the common postgres commands work the exact same way.

> I never faced issue like this. Maybe in some rolling release distro one can face this issue, but not with LTS releases. Anyway once you do apt update, it updates only package list. Once you've updated package list, apt install will always fetch newer package as per the list, whether you put it in another layer or not. It doesn't matter.

File: Dockerfile [hunk 1]
```diff
@@ -1,23 +1,20 @@
-From python:3.6
+From python:3.6-slim-stretch
 
 WORKDIR /usr/src/reminiscence
 
-COPY requirements.txt ./
+RUN apt-get update \
+  && apt-get install --no-install-recommends -y chromium netcat wget \
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply Dockerfile]: > update needs to be followed by upgrade.

No, this is wrong. It does not need to.

File: Dockerfile [hunk 2]
```diff
@@ -1,23 +1,20 @@
-From python:3.6
+From python:3.6-slim-stretch
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply Dockerfile]: Slim-stretch is 50MB https://hub.docker.com/_/python?tab=tags&page=1&name=slim-stretch and regular is 350MB!! https://hub.docker.com/_/python?tab=tags&page=1&name=3.6 .
The slim variant contains less bloat that is not required to run.
If you need any python modules you should put them in the requirements.txt and then they are going to be installed. 
I ran all my testing of reminiscence in this Docker image and it worked flawlessly. If you are concerned we can run all the tests with this PR.

[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply Dockerfile]: I ran both tests and they where successful.

File: Dockerfile [hunk 3]
```diff
@@ -1,23 +1,20 @@
...
 
 WORKDIR /usr/src/reminiscence
 
-COPY requirements.txt ./
+RUN apt-get update \
+  && apt-get install --no-install-recommends -y chromium netcat wget \
+  && wget https://github.com/wkhtmltopdf/wkhtmltopdf/releases/download/0.12.5/wkhtmltox_0.12.5-1.stretch_amd64.deb \
+  && apt-get install -y ./wkhtmltox_0.12.5-1.stretch_amd64.deb \
+  && rm ./wkhtmltox_0.12.5-1.stretch_amd64.deb \
+  && rm -rf /var/lib/apt/lists/*
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply Dockerfile]: See https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#run

> Official Debian and Ubuntu images automatically run apt-get clean, so explicit invocation is not required.

This is recommended in the best practices guide and also this removed the apt lists downloaded with ``apt update`` cause we don't need them anymore.

File: Dockerfile [hunk 4]
```diff
@@ -1,23 +1,20 @@
-From python:3.6
+From python:3.6-slim-stretch
 
 WORKDIR /usr/src/reminiscence
 
-COPY requirements.txt ./
+RUN apt-get update \
+  && apt-get install --no-install-recommends -y chromium netcat wget \
+  && wget https://github.com/wkhtmltopdf/wkhtmltopdf/releases/download/0.12.5/wkhtmltox_0.12.5-1.stretch_amd64.deb \
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply Dockerfile]: Then line 9 would be totally useless as we do not need or want the deb file in the final image. To achieve this we combine those 3 commands into one RUN that docker only adds the installed files of wkhtmltopdf to the layer.

File: Dockerfile [hunk 5]
```diff
@@ -1,23 +1,20 @@
...
 
-ADD . /usr/src/reminiscence
-
-RUN mkdir -p logs archive tmp
-
-RUN python manage.py applysettings --docker yes
+COPY . /usr/src/reminiscence
 
-RUN python manage.py generatesecretkey
+RUN mkdir -p logs archive tmp \
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply Dockerfile]: Which adds 3 unnecessary layers which probably override the same final multiple times. If you take a look at the offical Docker Hub images you can see that they quite heavily combine runs and sometimes only consist of one ADD and one RUN.

File: docker-compose.yml [hunk 6]
```diff
@@ -1,39 +1,30 @@
...
+      - ./logs:/usr/src/reminiscence/logs/
     depends_on:
-        - db
-        
+      - db
 
   db:
-    container_name: pg01
-    image: postgres:latest
+    image: postgres:11-alpine
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply docker-compose.yml]: Normally docker-compose generates the container name from the folder, service name and replication number. This makes it easy to know what a running container is and from where it came and how many of them there are.

No composer example I saw does name their container this way and it doesn't really add anything to the example. If I want to name my containers this way I can always do it.
An example should contain all the things necessary to run the container and no bloat.

**Comment by kanishka-linux** [comment]
Thanks for the link to the best practices in docker. It seems docker has different logic for updating/upgrading system compared to regular distro. Good to know that. 

All the points look valid except using `alpine postgres`. I don't see any reason in using it compared to regular image. If you can revert `alpine postgres`, I'm ready to accept PR.

Good Work!

File: Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (kanishka-linux) [reply Dockerfile]: I mean to say that it is considered as best practice in non-container regular system. But containers seems to have their own logic as per the link you have posted.

**Comment by SuperSandro2000** [comment]
I switched it back to the default postgres 11. I pinned the version because if you upgrade the container to a newer major version you usually need to dump and import your DB or otherwise the container won't start anymore.

**Pull request merged**